hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/TokenAirdropValidator.java

```diff
@@ -96,27 +95,37 @@ public void validateSemantics(
             @NonNull final ReadableNftStore nftStore) {
         var tokensConfig = context.configuration().getConfigData(TokensConfig.class);
         validateTrue(
-                op.tokenTransfers().size() <= tokensConfig.maxAllowedAirdropTransfersPerTx(), INVALID_TRANSACTION_BODY);
+                op.tokenTransfers().size() <= tokensConfig.maxAllowedAirdropTransfersPerTx(),
+                TOKEN_REFERENCE_LIST_SIZE_LIMIT_EXCEEDED);
 
         for (final var xfers : op.tokenTransfers()) {
             final var tokenId = xfers.tokenOrThrow();
-            validateTrue(tokenHasNoRoyaltyWithFallbackFee(tokenId, tokenStore), INVALID_TRANSACTION);
+            final var token = getIfUsable(tokenId, tokenStore);
+
             // process fungible token transfers if any.
             // PureChecks validates there is only one debit, so findFirst should return one item
             if (!xfers.transfers().isEmpty()) {
                 final var senderAccountAmount = xfers.transfers().stream()
                         .filter(item -> item.amount() < 0)
                         .findFirst();
                 final var senderId = senderAccountAmount.orElseThrow().accountIDOrThrow();
-                final var senderAccount = accountStore.getAliasedAccountById(senderId);
-                validateTrue(senderAccount != null, INVALID_ACCOUNT_ID);
-                // 1. Validate allowances and token associations
+                final var senderAccount =
+                        getIfUsableForAliasedId(senderId, accountStore, context.expiryValidator(), INVALID_ACCOUNT_ID);
+                // 1. Validate token associations
                 validateFungibleTransfers(
                         context.payer(), senderAccount, tokenId, senderAccountAmount.get(), tokenRelStore);
             }
 
             // process non-fungible tokens transfers if any
             if (!xfers.nftTransfers().isEmpty()) {
+                for (var transfer : xfers.nftTransfers()) {
+                    final var receiver = transfer.receiverAccountID();
+                    if (!isExemptFromCustomFees(token, receiver)) {
+                        validateTrue(
+                                tokenHasNoRoyaltyWithFallbackFee(token.tokenId(), tokenStore), INVALID_TRANSACTION);
```

Comment:
We should probably use `TOKEN_AIRDROP_WITH_FALLBACK_ROYALTY`

Reply:
agree, resolved